crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,696 @@
+    // more than the market can pay or selling less than the market wants)
+    // get flagged as liquidity orders. The reasoning is that these orders
+    // are not intended to be filled immediately and so need to be treated
+    // slightly differently by the protocol.
+    let is_liquidity_order =
+        if is_order_outside_market_price(&parameters.sell_amount, &parameters.buy_amount, &quote) {
+            tracing::debug!(%order_uid, ?owner, "order being flagged as outside market price");
+            true
+        } else {
+            false
+        };
+
+    let order = database::orders::Order {
+        uid: ByteArray(order_uid.0),
+        owner: ByteArray(owner.0),
+        creation_timestamp: chrono::offset::Utc::now(),
+        sell_token: ByteArray(order_data.sell_token.0),
+        buy_token: ByteArray(order_data.buy_token.0),
+        receiver: order_data.receiver.map(|h160| ByteArray(h160.0)),
+        sell_amount: u256_to_big_decimal(&order_data.sell_amount),
+        buy_amount: u256_to_big_decimal(&order_data.buy_amount),
+        valid_to: order_data.valid_to as i64,
+        app_data: ByteArray(order_data.app_data.0),
+        fee_amount: u256_to_big_decimal(&order_data.fee_amount),
+        kind: order_kind_into(order_data.kind),
+        partially_fillable: order_data.partially_fillable,
+        signature: order_placement.signature.1 .0.clone(),
+        signing_scheme: signing_scheme_into(signing_scheme),
+        settlement_contract: ByteArray(settlement_contract.0),
+        sell_token_balance: sell_token_source_into(order_data.sell_token_balance),
+        buy_token_balance: buy_token_destination_into(order_data.buy_token_balance),
+        full_fee_amount: u256_to_big_decimal(&full_fee_amount),
+        is_liquidity_order,
+        cancellation_timestamp: None,
+    };
+
+    let onchain_order_placement_event = OnchainOrderPlacement {
+        order_uid: ByteArray(order_uid.0),
+        sender: ByteArray(order_placement.sender.0),
+    };
+    Ok(Some((onchain_order_placement_event, order)))
+}
+
+fn convert_signature_data_to_owner(data: Vec<u8>) -> Result<H160> {
+    let owner = H160::from_slice(&data[..20]);
+    Ok(owner)
+}
+
+fn extract_order_data_from_onchain_order_placement_event(
+    order_placement: &ContractOrderPlacement,
+    domain_separator: DomainSeparator,
+) -> Result<(OrderData, H160, SigningScheme, OrderUid)> {
+    let (signing_scheme, owner) = match order_placement.signature.0 {
+        0 => (
+            SigningScheme::Eip1271,
+            convert_signature_data_to_owner(order_placement.signature.1 .0.clone())?,
+        ),
+        1 => (SigningScheme::PreSign, order_placement.sender),
+        _ => bail!("unreachable state while parsing owner"),
+    };
+
+    let order_data = OrderData {
+        sell_token: order_placement.order.0,
+        buy_token: order_placement.order.1,
+        receiver: Some(order_placement.order.2),
```

Comment:
[2] For ETH flow it doesn't matter, but for user-created onchain order the receiver could be zero. I believe that `Some(0x0...0)` should be invalid and the reason why this is an `Option` in the first place.

Reply:
Very good catch! nice